mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/DeleteByQueryPerformer.java

```diff
@@ -0,0 +1,105 @@
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.elasticsearch;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.elasticsearch.action.ListenableActionFuture;
+import org.elasticsearch.action.bulk.BulkRequestBuilder;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.msgpack.core.Preconditions;
+
+public class DeleteByQueryPerformer {
+
+    public static final String DELETE_BY_QUERY_POOL_SIZE = "DeleteByQueryPoolSize";
+    public static final String DELETE_BY_QUERY_BATCH_SIZE = "DeleteByQueryBatchSize";
+    private final ClientProvider clientProvider;
+    private final ExecutorService executor;
+    private final int batchSize;
+
+    @Inject
+    public DeleteByQueryPerformer(ClientProvider clientProvider,
+                                  @Named(DELETE_BY_QUERY_POOL_SIZE) Integer threadPoolSize,
+                                  @Named(DELETE_BY_QUERY_BATCH_SIZE) Integer batchSize) {
+        checkValueIsPositive(threadPoolSize);
+        checkValueIsPositive(batchSize);
+        this.clientProvider = clientProvider;
+        this.executor = Executors.newFixedThreadPool(threadPoolSize);
+        this.batchSize = batchSize;
+    }
+
+    private void checkValueIsPositive(Integer value) {
+        Preconditions.checkNotNull(value);
+        Preconditions.checkArgument(value > 0, "value must be positive");
+    }
+
+    @PreDestroy
+    public void stop() {
+        executor.shutdownNow();
+    }
+
+    public void deleteByQuery(QueryBuilder queryBuilder) {
+        executor.execute(() -> doDeleteByQuery(queryBuilder));
+    }
+
+    private void doDeleteByQuery(QueryBuilder queryBuilder) {
+        try (Client client = clientProvider.get()) {
+            ListenableActionFuture<SearchResponse> searchResponseFuture = client.prepareSearch(ElasticSearchIndexer.MAILBOX_INDEX)
+                .setTypes(ElasticSearchIndexer.MESSAGE_TYPE)
+                .setScroll(new TimeValue(60000))
+                .setQuery(queryBuilder)
+                .setSize(batchSize)
+                .execute();
+
+            while(true) {
+                SearchResponse searchResponse = searchResponseFuture.actionGet();
+                if (searchResponse.getHits().getHits().length == 0) {
+                    break;
+                }
+                searchResponseFuture = client.prepareSearchScroll(searchResponse.getScrollId())
```

Comment:
This code was taken from https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/java-search-scrolling.html

I reworked it to use future. But sure, I will completly rework it. And I should have done that first.